libtsuba/test/storage-format-version/v5-optional-datastructure-rdklsh-index.cpp

```diff
@@ -0,0 +1,129 @@
+  for (size_t i = 0; i < 128; i++) {
+    for (size_t j = 0; j < 64; j++) {
+      for (size_t k = 0; k < 32; k++) {
+        hashes[i][j][k] = i + j + k;
+      }
+    }
+  }
+  return hashes;
+}
+
+std::vector<katana::DynamicBitset>
+GenerateFingerprints() {
+  std::vector<katana::DynamicBitset> fingerprints;
+
+  for (size_t i = 0; i < 3; i++) {
+    katana::DynamicBitset bset;
+    for (size_t j = 0; j < i; j++) {
+      bset.resize(j + 1);
+      bset.set(j);
+    }
+    fingerprints.emplace_back(std::move(bset));
+  }
+
+  return fingerprints;
+}
+
+std::vector<std::string>
+GenerateSmiles() {
+  std::vector<std::string> smiles = {"smile1", "smile2", "smile3", "smile4"};
+  return smiles;
+}
+
+katana::RDKLSHIndex
+GenerateIndex() {
+  katana::RDKLSHIndex index;
+  //tODO(emcginnis) test that we can store/load these files
+
+  std::vector<katana::DynamicBitset> fingerprints = GenerateFingerprints();
+  index.set_num_hashes_per_bucket(16);
+  index.set_num_buckets(96);
+  index.set_fingerprint_length(42);
+  index.set_num_fingerprints(fingerprints.size());
+  index.set_hash_structure(GenerateHashes());
+  index.set_fingerprints(std::move(fingerprints));
+  index.set_smiles(GenerateSmiles());
+  return index;
+}
+
+void
+ValidateIndex(katana::RDKLSHIndex& index) {
+  KATANA_LOG_ASSERT(index.num_hashes_per_bucket() == 16);
+  KATANA_LOG_ASSERT(index.num_buckets() == 96);
+  KATANA_LOG_ASSERT(index.fingerprint_length() == 42);
+  KATANA_LOG_ASSERT(index.num_fingerprints() == 3);
+  KATANA_LOG_ASSERT(index.hash_structure() == GenerateHashes());
+  KATANA_LOG_ASSERT(index.fingerprints() == GenerateFingerprints());
+  KATANA_LOG_ASSERT(index.smiles() == GenerateSmiles());
+}
+
+katana::Result<void>
+TestRoundTripRDKLSHIndex(const std::string& rdg_dir) {
+  KATANA_LOG_ASSERT(!rdg_dir.empty());
+  katana::RDKLSHIndex index = GenerateIndex();
+  ValidateIndex(index);
+
+  katana::RDG rdg = KATANA_CHECKED(LoadRDG(rdg_dir));
+  KATANA_CHECKED(rdg.WriteRDKLSHIndex(index));
+
+  // Read the index back and ensure it matches what we put in
+  katana::RDKLSHIndex index_2 = KATANA_CHECKED(rdg.LoadRDKLSHIndex());
+  ValidateIndex(index_2);
+
+  // Store/Load the RDG in a new location and ensure our index is still intact
+  std::string rdg_dir2 = KATANA_CHECKED(WriteRDG(std::move(rdg)));
+  katana::RDG rdg2 = KATANA_CHECKED(LoadRDG(rdg_dir2));
+  katana::RDKLSHIndex index_3 = KATANA_CHECKED(rdg2.LoadRDKLSHIndex());
+  ValidateIndex(index_3);
+
+  return katana::ResultSuccess();
+}
+
+int
+main(int argc, char* argv[]) {
+  if (auto init_good = katana::InitTsuba(); !init_good) {
+    KATANA_LOG_FATAL("katana::InitTsuba: {}", init_good.error());
+  }
+  katana::GaloisRuntime Katana_runtime;
+
+  if (argc <= 1) {
+    KATANA_LOG_FATAL("missing rdg file directory");
+  }
+
+  // Ensure the feature flag is actually set
+  KATANA_LOG_ASSERT(KATANA_EXPERIMENTAL_ENABLED(UnstableRDGStorageFormat));
```

Comment:
Maybe I am getting my self confused -- but just to make sure --are we requiring that users of this RDG optional structure feature will require someone to enable experimental RDG features or are we saying that users of the LSH Index stored in the RDG will require this flag? This is just until we get to the point where we commit to supporting this long-term -- right?

Reply:
yep, the is only while we are stabilizing what the RDKLSHIndex looks like to avoid needless tinkering with the rdg_storage_format